src/rest_vol_dataset.c

```diff
@@ -5217,6 +5369,370 @@ RV_get_cmpd_subset_nmembers(hid_t src_type_id, hid_t dst_type_id, size_t *num_cm
+        out_buf_curr_pos += vl.len * parent_dtype_size;
+    }
+
+    *out = out_buf;
+done:
+    if (ret_value < 0 && out_buf)
+        RV_free(out_buf);
+
+    return ret_value;
+}
+
+/*-------------------------------------------------------------------------
+ * Function:    RV_unpack_vlen_data
+ *
+ * Purpose:     Converts a buffer of packed hvl_t instances to a buffer of
+ *              hvl_t instances.
+ *
+ *              Allocates memory for hvl_t instances which should be freed with
+ *              H5Treclaim, and allocates under *out which should
+ *              be freed directly by the calling function.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ * Programmer:  Matthew Larson
+ *              March, 2024
+ */
+static herr_t
+RV_unpack_vlen_data(char *in, hid_t vlen_dtype_id, size_t nelems, void **out)
+{
+    herr_t ret_value = SUCCEED;
+    char  *out_buf   = NULL;
+    char  *out_buf_curr_ptr;
+    char  *in_buf_ptr   = in;
+    hid_t  parent_dtype = H5I_INVALID_HID;
+    size_t parent_dtype_size;
+    hvl_t *vl;
+
+    if (!in)
+        FUNC_GOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "provided vlen datatype was NULL");
+
+    if (!out)
+        FUNC_GOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "provided output pointer was NULL");
+
+    if ((out_buf = RV_calloc(nelems * sizeof(hvl_t))) == NULL)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "can't allocate space for unpacking vlen data");
+
+    if ((parent_dtype = H5Tget_super(vlen_dtype_id)) < 0)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get parent of vlen datatype");
+
+    if ((parent_dtype_size = H5Tget_size(parent_dtype)) == 0)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get size of vlen parent datatype");
+
+    for (size_t i = 0; i < nelems; i++) {
+        vl = ((hvl_t *)out_buf) + i;
+        uint32_t num_seq_bytes;
+
+        memcpy(&num_seq_bytes, in_buf_ptr, sizeof(uint32_t));
+        in_buf_ptr += sizeof(uint32_t);
+
+        if (num_seq_bytes % parent_dtype_size != 0)
+            FUNC_GOTO_ERROR(H5E_DATASET, H5E_BADVALUE, FAIL,
+                            "vlen sequence length is not a multiple of element size");
+
+        vl->len = num_seq_bytes / parent_dtype_size;
+
+        /* Allocate memory for and copy sequence data */
+        if (vl->len > 0) {
+            if ((vl->p = RV_calloc(num_seq_bytes)) == NULL)
+                FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL,
+                                "can't allocate space for unpacking vlen data");
+
+            memcpy(vl->p, in_buf_ptr, num_seq_bytes);
+        }
+        else {
+            vl->p = NULL;
+        }
+
+        in_buf_ptr += num_seq_bytes;
+    }
+
+    *out = out_buf;
+
+done:
+
+    if (ret_value < 0 && out_buf) {
+        /* Free any memory allocated for individual sequences */
+        for (size_t i = 0; i < nelems; i++) {
+            vl = ((hvl_t *)out_buf) + i;
```

Comment:
Casting `out_buf` to `hvl_t *` here is not correct due to alignment requirements. You should be able to find examples of this elsewhere, but the proper way is to have a local `hvl_t` stack variable that you memcpy into from the appropriate index into `out_buf` (as a `char *`) and then perform your free operations on that variable's `.p` field.

In this case no modifications to the buffer need to be made since it was allocated locally, but if you needed to modify something in the original buffer (such as when setting `vl->p` to NULL here), you'd memcpy from the stack variable back into the buffer at the same index and then move to the next element index.